First comp ver

Grillo-0 opened the pull request.

Grillo-0 committed:
Make the first compiler version

Grillo-0 committed:
src/: Rename call_args to just args

We only have one type of args, so just name them `args` instead of
`call_args`

Grillo-0 committed:
src/dry_bison.c: Start accepting blocks for assignments

The vinum description, says that we must accept blocks inside variabel
assignments. So do it.

Grillo-0 committed:
src/dry_bison: Accept blocks as symbols

Our language can accept blocks as variable names, this is done to have
meta-programming. So do it.

Grillo-0 committed:
src/dry_flex.l: Add '\r' as a non printable character

To support Windows new lines.

Grillo-0 committed:
src/ast.c: Make the assert on ast_node_add_child() non inclusive

As we use `num_childs` as the index to the node to be added, the assert
must be non inclusive to not cause any bugs.

Grillo-0 committed:
src/dry_bison.y: Accept ':' inside a TEXT node

By the language description, we must accept ':' inside a `TEXT` node. Do
that by changing the bison BNF of the text.

Grillo-0 committed:
Add a more complex test

We need a more complex .vin file that represents a real use of the
language.

Grillo-0 committed:
src/dry_flex.l: Make the WORD definition utf friendly

The [[:graph:]] group only accepts ASCII chars. Change the `WORD` rule
to be any character besides the `[`, `]` `:` and the non-printable ones.

Grillo-0 committed:
Add 'configure~' to .gitignore

Grillo-0 committed:
Fix AM_CFLAGS placement and adjust configure.ac to prevent unwanted flags

The AM_CFLAGS definition is on the wrong place, it should be on
src/Makefile.am, instead of Makefile.am. Also configure.ac is adding
some flags on it's own.

To solve this, move the AM_CFLGAS to the correct place and add make
CFLAGS empty before setting AC_PROG_CC on configure.ac. After that,
remove all warnings in the code.

Grillo-0 committed:
src/ast: Create a function to copy a node

In the future, wen evaluating the AST, we will need a function to make a
copy of a node, copying it's children recursively.

Grillo-0 committed:
src: Start evaluating the AST

Grillo-0 committed:
vinumc: Start to recognize '$*' as a reference to all args

The language has a way to reference all the args in a function
definition with '$*'.

As we can only decide if a '$*' is a arg reference by the location of
the token, the lexer returns the token `WORD` but it create a node with
the correct token `ARG_REF_ALL_ARGS`.

Grillo-0 committed:
src/ast.h: Include bison generated header

Grillo-0 committed:
src/vec.h: Create a dynamic array data structure

Add a polymorphic dynamic array data type by creating a macro that can
put elements in array represented by any struct that has the attributes
below:

- `base`: base pointer were the array is located, this pointer can be of
  any type.
- `len`: number of elements in the array
- `capacity`: number of elements that the array can hold

Also, add a helper macro to create those structs for a specific type.

MurilloMMartins committed:
src/dry_bison.y: make symbols case-insensitive

This change ensures that symbols are treated as case-insensitive as
described in the language specification. All symbol texts are now
converted to lowercase, including Unicode characters like 'Á' that were
previously unsupported.

MurilloMMartins committed:
src/vinumc.c: set locale to unicode characters

Add locale to support unicode character operations.

Grillo-0 committed:
src/ast: Create a function to print the AST in the dot language

I think it would be useful for visualising the AST if we could print it
in the dot language[1]. With this we could use the output with the `dot`
utility to generate the visualisation.

To do this, create a new function `ast_dot()` that prints the AST to a
`FILE*` in the dot language.

[1]: https://en.wikipedia.org/wiki/DOT_(graph_description_language)

Grillo-0 committed:
src: Create an utility to print the AST of a program in dot format

Use the new function `ast_dot()` to create the utility `vin2dot` to
print to stdout the dot representation[1] of program's AST.

[1]: https://en.wikipedia.org/wiki/DOT_(graph_description_language)

Grillo-0 committed:
src/ast.c: Move `ast_copy_node()` to before `ast_print()`

Currently, the function `ast_copy_node()` is in the middle of the helper
functions that implement `ast_print()`.

Move it to before those functions and after `ast_add_node()`, like it is
on `src/ast.h`

Grillo-0 committed:
src/ast: Use a vec for storing the child indices on the `ast_node`

The MAX_AST_NODE_CHILDS limit was useful for prototyping the initial
compiler. Now it would be better to not have this limit.

Let's solve this by using a vec instead of just an array with a fixed
length.

Grillo-0 committed:
src/vec.h: Add VEC_RESERVE macro

In some cases, when we know the number of elements that are going to be
added, it's better to reserve this capacity beforehand, to avoid
unnecessary `realloc` calls.

Grillo-0 committed:
src/ast: Use a vec for storing ast_nodes on the ast

The MAX_AST_NODES limit was useful for prototyping the initial compiler.
Now it would be better to not have this limit.

Let's solve this by using a vec instead of just an array with a fixed
length.

Grillo-0 committed:
src/vec.h: Add a VEC_AT macro

Add this macro to abstract more the vec, so you don't have to know about
the `base` field.

Grillo-0 committed:
Merge pull request #10 from vinumdoc/no-limit-ast

Remove the hardcoded limits on the AST

Grillo-0 committed:
Merge pull request #9 from vinumdoc/vin2dot

Vin2dot

Grillo-0 committed:
Merge pull request #7 from vinumdoc/3-make-symbol-case-insensitive

Make symbol case insensitive

artP2 commented on out.html:
Add this file to .gitignore to exclude it from version control.

artP2 commented on out.txt:
Add this file to .gitignore to exclude it from version control.

Grillo-0 committed:
vinumc: Remove the vinumc generated files

On a couple of commits the vinumc generated files are present, this
should not have happened, as it makes no sense to have them among the
project files.

To fix that, remove those files and add them to the gitignore.
